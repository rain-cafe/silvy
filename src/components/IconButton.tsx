import { IconType } from 'react-icons';
import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import {Button} from './Button';
import styles from './IconButton.module.scss';
import { classnames } from '../utils/classnames';
import { Animations } from '../constants/components';

export interface IconButtonOwnProps {
    icon: IconType;
    hoverIcon: IconType;
    size?: number;
    className?: string;
    disabled?: boolean;
}

export type IconButtonProps<
    T extends React.ElementType
> = PolymorphicPropsWithoutRef<IconButtonOwnProps, T>;

export function IconButton<E extends React.ElementType>({
    as,
    icon: Icon,
    hoverIcon: HoverIcon,
    size = 30,
    className,
    disabled = false,
    ...extraProps
}: IconButtonProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as;
    const props = {
        ...extraProps,
        className: classnames(
            styles.iconButton,
            disabled && styles.disabled,
            className
        )
    };

    return (
        <Button as={Element} animation={Animations.NONE} {...props}>
            <Icon className={styles.icon} size={size} />
            <HoverIcon className={styles.hoverIcon} size={size} />
        </Button>
    );
}