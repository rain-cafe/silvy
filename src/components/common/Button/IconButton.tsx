import { IconType } from 'react-icons';
import type { PolymorphicPropsWithoutRef } from 'react-polymorphic-types';
import classNames from 'classnames';
import {Button, ButtonOwnProps} from './Button';
import styles from './IconButton.module.scss';
import { Animations } from '../../../constants/components';
import { useReadOnlyCachedState } from '../../hooks/use-cached-state';

export interface IconButtonOwnProps extends Omit<ButtonOwnProps, 'children'> {
    icon: IconType;
    hoverIcon?: IconType;
    children?: React.ReactNode;
    className?: string;
    disabled?: boolean;
};

export type IconButtonProps<
    T extends React.ElementType
> = PolymorphicPropsWithoutRef<IconButtonOwnProps, T>;

export function IconButton<E extends React.ElementType>({
    as,
    icon: Icon,
    hoverIcon: HoverIcon,
    children,
    className: externalClassName,
    ...extraProps
}: IconButtonProps<E>) {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const Element: React.ElementType = as;
    const className = useReadOnlyCachedState(() => {
        return classNames(
            styles.iconButton,
            externalClassName,
            {
                [styles.hasHoverIcon]: HoverIcon
            }
        );
    }, [HoverIcon, externalClassName]);

    return (
        <Button
            data-testid='icon-button'
            {...extraProps}
            as={Element}
            animation={HoverIcon ? Animations.NONE : Animations.DARKEN}
            className={className}
        >
            <Icon className={styles.icon} size={30} />
            {HoverIcon && (
                <HoverIcon className={styles.hoverIcon} size={30} />
            )}
            {children}
        </Button>
    );
}