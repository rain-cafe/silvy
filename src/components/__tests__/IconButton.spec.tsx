import Link from 'next/link';
import React from 'react';
import { BsPatchCheck, BsPatchCheckFill } from 'react-icons/bs';
import { createRenderComponent } from '../../../__tests__/render';
import { IconButton, IconButtonProps } from '../IconButton';

describe('Component(Button)', () => {
    const renderButton = createRenderComponent<IconButtonProps<React.ElementType>>(IconButton, {
        icon: BsPatchCheck,
        hoverIcon: BsPatchCheckFill
    });

    it('should display the Button', () => {
        const { baseElement } = renderButton()

        expect(baseElement).toMatchSnapshot()
    });

    it('should support overriding the tag type', () => {
        const { baseElement } = renderButton({
            as: 'a'
        });

        expect(baseElement).toMatchSnapshot()
    });

    // TODO: Figure out how to get proper type checking working with Poly Components
    it('should support NextJS Links', () => {
        const { baseElement } = renderButton({
            as: Link,
            href: 'https://google.com'
        } as any);

        expect(baseElement).toMatchSnapshot()
    });

    it('should support changing the size', () => {
        const { baseElement } = renderButton({
            size: 30
        });

        expect(baseElement).toMatchSnapshot()
    });

    it('should support being disabled', () => {
        const { baseElement } = renderButton({
            disabled: true
        });

        expect(baseElement).toMatchSnapshot()
    });
});
