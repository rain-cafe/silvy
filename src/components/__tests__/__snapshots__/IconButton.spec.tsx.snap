// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Button) should display the Button 1`] = `
<body>
  <div>
    <button
      class="button sz-small an-none iconButton"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.354 6.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
        <path
          d="m10.273 2.513-.921-.944.715-.698.622.637.89-.011a2.89 2.89 0 0 1 2.924 2.924l-.01.89.636.622a2.89 2.89 0 0 1 0 4.134l-.637.622.011.89a2.89 2.89 0 0 1-2.924 2.924l-.89-.01-.622.636a2.89 2.89 0 0 1-4.134 0l-.622-.637-.89.011a2.89 2.89 0 0 1-2.924-2.924l.01-.89-.636-.622a2.89 2.89 0 0 1 0-4.134l.637-.622-.011-.89a2.89 2.89 0 0 1 2.924-2.924l.89.01.622-.636a2.89 2.89 0 0 1 4.134 0l-.715.698a1.89 1.89 0 0 0-2.704 0l-.92.944-1.32-.016a1.89 1.89 0 0 0-1.911 1.912l.016 1.318-.944.921a1.89 1.89 0 0 0 0 2.704l.944.92-.016 1.32a1.89 1.89 0 0 0 1.912 1.911l1.318-.016.921.944a1.89 1.89 0 0 0 2.704 0l.92-.944 1.32.016a1.89 1.89 0 0 0 1.911-1.912l-.016-1.318.944-.921a1.89 1.89 0 0 0 0-2.704l-.944-.92.016-1.32a1.89 1.89 0 0 0-1.912-1.911l-1.318.016z"
        />
      </svg>
      <svg
        class="hoverIcon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.067.87a2.89 2.89 0 0 0-4.134 0l-.622.638-.89-.011a2.89 2.89 0 0 0-2.924 2.924l.01.89-.636.622a2.89 2.89 0 0 0 0 4.134l.637.622-.011.89a2.89 2.89 0 0 0 2.924 2.924l.89-.01.622.636a2.89 2.89 0 0 0 4.134 0l.622-.637.89.011a2.89 2.89 0 0 0 2.924-2.924l-.01-.89.636-.622a2.89 2.89 0 0 0 0-4.134l-.637-.622.011-.89a2.89 2.89 0 0 0-2.924-2.924l-.89.01-.622-.636zm.287 5.984-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708.708z"
        />
      </svg>
    </button>
  </div>
</body>
`;

exports[`Component(Button) should support NextJS Links 1`] = `
<body>
  <div>
    <a
      class="button sz-small an-none iconButton"
      href="https://google.com"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.354 6.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
        <path
          d="m10.273 2.513-.921-.944.715-.698.622.637.89-.011a2.89 2.89 0 0 1 2.924 2.924l-.01.89.636.622a2.89 2.89 0 0 1 0 4.134l-.637.622.011.89a2.89 2.89 0 0 1-2.924 2.924l-.89-.01-.622.636a2.89 2.89 0 0 1-4.134 0l-.622-.637-.89.011a2.89 2.89 0 0 1-2.924-2.924l.01-.89-.636-.622a2.89 2.89 0 0 1 0-4.134l.637-.622-.011-.89a2.89 2.89 0 0 1 2.924-2.924l.89.01.622-.636a2.89 2.89 0 0 1 4.134 0l-.715.698a1.89 1.89 0 0 0-2.704 0l-.92.944-1.32-.016a1.89 1.89 0 0 0-1.911 1.912l.016 1.318-.944.921a1.89 1.89 0 0 0 0 2.704l.944.92-.016 1.32a1.89 1.89 0 0 0 1.912 1.911l1.318-.016.921.944a1.89 1.89 0 0 0 2.704 0l.92-.944 1.32.016a1.89 1.89 0 0 0 1.911-1.912l-.016-1.318.944-.921a1.89 1.89 0 0 0 0-2.704l-.944-.92.016-1.32a1.89 1.89 0 0 0-1.912-1.911l-1.318.016z"
        />
      </svg>
      <svg
        class="hoverIcon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.067.87a2.89 2.89 0 0 0-4.134 0l-.622.638-.89-.011a2.89 2.89 0 0 0-2.924 2.924l.01.89-.636.622a2.89 2.89 0 0 0 0 4.134l.637.622-.011.89a2.89 2.89 0 0 0 2.924 2.924l.89-.01.622.636a2.89 2.89 0 0 0 4.134 0l.622-.637.89.011a2.89 2.89 0 0 0 2.924-2.924l-.01-.89.636-.622a2.89 2.89 0 0 0 0-4.134l-.637-.622.011-.89a2.89 2.89 0 0 0-2.924-2.924l-.89.01-.622-.636zm.287 5.984-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708.708z"
        />
      </svg>
    </a>
  </div>
</body>
`;

exports[`Component(Button) should support being disabled 1`] = `
<body>
  <div>
    <button
      class="button sz-small an-none iconButton disabled"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.354 6.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
        <path
          d="m10.273 2.513-.921-.944.715-.698.622.637.89-.011a2.89 2.89 0 0 1 2.924 2.924l-.01.89.636.622a2.89 2.89 0 0 1 0 4.134l-.637.622.011.89a2.89 2.89 0 0 1-2.924 2.924l-.89-.01-.622.636a2.89 2.89 0 0 1-4.134 0l-.622-.637-.89.011a2.89 2.89 0 0 1-2.924-2.924l.01-.89-.636-.622a2.89 2.89 0 0 1 0-4.134l.637-.622-.011-.89a2.89 2.89 0 0 1 2.924-2.924l.89.01.622-.636a2.89 2.89 0 0 1 4.134 0l-.715.698a1.89 1.89 0 0 0-2.704 0l-.92.944-1.32-.016a1.89 1.89 0 0 0-1.911 1.912l.016 1.318-.944.921a1.89 1.89 0 0 0 0 2.704l.944.92-.016 1.32a1.89 1.89 0 0 0 1.912 1.911l1.318-.016.921.944a1.89 1.89 0 0 0 2.704 0l.92-.944 1.32.016a1.89 1.89 0 0 0 1.911-1.912l-.016-1.318.944-.921a1.89 1.89 0 0 0 0-2.704l-.944-.92.016-1.32a1.89 1.89 0 0 0-1.912-1.911l-1.318.016z"
        />
      </svg>
      <svg
        class="hoverIcon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.067.87a2.89 2.89 0 0 0-4.134 0l-.622.638-.89-.011a2.89 2.89 0 0 0-2.924 2.924l.01.89-.636.622a2.89 2.89 0 0 0 0 4.134l.637.622-.011.89a2.89 2.89 0 0 0 2.924 2.924l.89-.01.622.636a2.89 2.89 0 0 0 4.134 0l.622-.637.89.011a2.89 2.89 0 0 0 2.924-2.924l-.01-.89.636-.622a2.89 2.89 0 0 0 0-4.134l-.637-.622.011-.89a2.89 2.89 0 0 0-2.924-2.924l-.89.01-.622-.636zm.287 5.984-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708.708z"
        />
      </svg>
    </button>
  </div>
</body>
`;

exports[`Component(Button) should support changing the size 1`] = `
<body>
  <div>
    <button
      class="button sz-small an-none iconButton"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.354 6.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
        <path
          d="m10.273 2.513-.921-.944.715-.698.622.637.89-.011a2.89 2.89 0 0 1 2.924 2.924l-.01.89.636.622a2.89 2.89 0 0 1 0 4.134l-.637.622.011.89a2.89 2.89 0 0 1-2.924 2.924l-.89-.01-.622.636a2.89 2.89 0 0 1-4.134 0l-.622-.637-.89.011a2.89 2.89 0 0 1-2.924-2.924l.01-.89-.636-.622a2.89 2.89 0 0 1 0-4.134l.637-.622-.011-.89a2.89 2.89 0 0 1 2.924-2.924l.89.01.622-.636a2.89 2.89 0 0 1 4.134 0l-.715.698a1.89 1.89 0 0 0-2.704 0l-.92.944-1.32-.016a1.89 1.89 0 0 0-1.911 1.912l.016 1.318-.944.921a1.89 1.89 0 0 0 0 2.704l.944.92-.016 1.32a1.89 1.89 0 0 0 1.912 1.911l1.318-.016.921.944a1.89 1.89 0 0 0 2.704 0l.92-.944 1.32.016a1.89 1.89 0 0 0 1.911-1.912l-.016-1.318.944-.921a1.89 1.89 0 0 0 0-2.704l-.944-.92.016-1.32a1.89 1.89 0 0 0-1.912-1.911l-1.318.016z"
        />
      </svg>
      <svg
        class="hoverIcon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.067.87a2.89 2.89 0 0 0-4.134 0l-.622.638-.89-.011a2.89 2.89 0 0 0-2.924 2.924l.01.89-.636.622a2.89 2.89 0 0 0 0 4.134l.637.622-.011.89a2.89 2.89 0 0 0 2.924 2.924l.89-.01.622.636a2.89 2.89 0 0 0 4.134 0l.622-.637.89.011a2.89 2.89 0 0 0 2.924-2.924l-.01-.89.636-.622a2.89 2.89 0 0 0 0-4.134l-.637-.622.011-.89a2.89 2.89 0 0 0-2.924-2.924l-.89.01-.622-.636zm.287 5.984-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708.708z"
        />
      </svg>
    </button>
  </div>
</body>
`;

exports[`Component(Button) should support overriding the tag type 1`] = `
<body>
  <div>
    <a
      class="button sz-small an-none iconButton"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.354 6.146a.5.5 0 0 1 0 .708l-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708 0z"
          fill-rule="evenodd"
        />
        <path
          d="m10.273 2.513-.921-.944.715-.698.622.637.89-.011a2.89 2.89 0 0 1 2.924 2.924l-.01.89.636.622a2.89 2.89 0 0 1 0 4.134l-.637.622.011.89a2.89 2.89 0 0 1-2.924 2.924l-.89-.01-.622.636a2.89 2.89 0 0 1-4.134 0l-.622-.637-.89.011a2.89 2.89 0 0 1-2.924-2.924l.01-.89-.636-.622a2.89 2.89 0 0 1 0-4.134l.637-.622-.011-.89a2.89 2.89 0 0 1 2.924-2.924l.89.01.622-.636a2.89 2.89 0 0 1 4.134 0l-.715.698a1.89 1.89 0 0 0-2.704 0l-.92.944-1.32-.016a1.89 1.89 0 0 0-1.911 1.912l.016 1.318-.944.921a1.89 1.89 0 0 0 0 2.704l.944.92-.016 1.32a1.89 1.89 0 0 0 1.912 1.911l1.318-.016.921.944a1.89 1.89 0 0 0 2.704 0l.92-.944 1.32.016a1.89 1.89 0 0 0 1.911-1.912l-.016-1.318.944-.921a1.89 1.89 0 0 0 0-2.704l-.944-.92.016-1.32a1.89 1.89 0 0 0-1.912-1.911l-1.318.016z"
        />
      </svg>
      <svg
        class="hoverIcon"
        fill="currentColor"
        height="30"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="30"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.067.87a2.89 2.89 0 0 0-4.134 0l-.622.638-.89-.011a2.89 2.89 0 0 0-2.924 2.924l.01.89-.636.622a2.89 2.89 0 0 0 0 4.134l.637.622-.011.89a2.89 2.89 0 0 0 2.924 2.924l.89-.01.622.636a2.89 2.89 0 0 0 4.134 0l.622-.637.89.011a2.89 2.89 0 0 0 2.924-2.924l-.01-.89.636-.622a2.89 2.89 0 0 0 0-4.134l-.637-.622.011-.89a2.89 2.89 0 0 0-2.924-2.924l-.89.01-.622-.636zm.287 5.984-3 3a.5.5 0 0 1-.708 0l-1.5-1.5a.5.5 0 1 1 .708-.708L7 8.793l2.646-2.647a.5.5 0 0 1 .708.708z"
        />
      </svg>
    </a>
  </div>
</body>
`;
